src/Nito.AsyncEx.Coordination/AsyncLazy.cs

```diff
@@ -121,17 +115,18 @@ public Task<T> Task
 
         private Func<Task<T>> RetryOnFailure(Func<Task<T>> factory)
         {
-            return async () =>
+            var originalFactory = factory;
+            return factory = async () =>
             {
                 try
                 {
-                    return await factory().ConfigureAwait(false);
+                    return await originalFactory().ConfigureAwait(false);
                 }
                 catch
                 {
                     lock (_mutex)
                     {
-                        _instance = new Lazy<Task<T>>(_factory);
+                        _instance = new Lazy<Task<T>>(factory);
```

Comment:
Doesn't this both capture the new factory variable and effectively create a new delegate each time it fails? I'm not 100% on this codebase, but it doesn't feel right to me - perhaps 
```cs
_instance = new Lazy<Task<T>>(RetryOnFailure(originalFactory));
```
Please ignore if I am not following this perfectly, as I am well aware this is black magic.

Reply:
No, the delegate is only created once. It's re-using the captured delegate. What you have there, calling the `RetryOnFailure` function again, would be creating a new delegate every time.

Reply:
@timcassell Okay cool - as I said; I am not 100% on how this exactly works, but didn't like the fact it looks like it captures the delegate variable inside the delegate. It felt like this would nest the delegate inside itself on each failure but looking again it seems like you are right.

Reply:
It technically does create a circular reference, where the delegate stores a reference to a class that stores a reference to itself. But that circular reference is only created once.